README
NAME
    Kavorka - function signatures with the lure of the animal

SYNOPSIS
       use Kavorka;
   
       fun maxnum (Num @numbers) {
          my $max = shift @numbers;
          for (@numbers) {
             $max = $_ if $max < $_;
          }
          return $max;
       }
   
       my $biggest = maxnum(42, 3.14159, 666);

STATUS
    Kavorka is still at a very early stage of development; there are likely to
    be many bugs that still need to be shaken out. Certain syntax features are
    a little odd and may need to be changed in incompatible ways.

DESCRIPTION
    Kavorka provides `fun` and `method` keywords for declaring functions and
    methods. It uses Perl 5.14's keyword API, so should work more reliably
    than source filters or Devel::Declare-based modules.

    The syntax provided by Kavorka is largely inspired by Perl 6, though it
    has also been greatly influenced by Method::Signatures and
    Function::Parameters.

    For information using the keywords exported by Kavorka:

    *   Kavorka::Manual::Functions

    *   Kavorka::Manual::Methods

    *   Kavorka::Manual::MethodModifiers

    *   Kavorka::Manual::MultiSubs

  Exports
    `-default`
        Exports `fun` and `method`.

    `-modifiers`
        Exports `before`, `after`, and `around`.

    `-allmodifiers`
        Exports `before`, `after`, `around`, `augment`, and `override`.

    `-all`
        Exports `fun`, `method`, `before`, `after`, `around`, `augment`,
        `override`, `classmethod`, `objectmethod`, and `multi`.

    For example:

       # Everything except objectmethod and multi...
       use Kavorka qw( -default -allmodifiers classmethod );

    You can rename imported functions:

       use Kavorka method => { -as => 'meth' };

    You can provide alternative implementations:

       # use My::Sub::Method instead of Kavorka::Sub::Method
       use Kavorka method => { implementation => 'My::Sub::Method' };

    Or add traits to the default implementation:

       use Kavorka method => { traits => ['My::Sub::Role::Foo'] };

    See Exporter::Tiny for more tips.

  Function Introspection API
    The coderef for any sub created by Kavorka can be passed to the
    `Kavorka->info` method. This returns a blessed object that does the
    Kavorka::Sub role.

       fun foo (:$x, :$y) { }
   
       my $info = Kavorka->info(\&foo);
   
       my $function_name = $info->qualified_name;
       my @named_params  = $info->signature->named_params;
   
       say $named_params[0]->named_names->[0];   # says 'x'

    See Kavorka::Sub, Kavorka::Signature and Kavorka::Parameter for further
    details.

    If you're using Moose, consider using MooseX::KavorkaInfo to expose
    Kavorka method signatures via the meta object protocol.

    Kavorka::Manual::API provides more details and examples using the
    introspection API.

CAVEATS
    *   As noted in Kavorka::Manual::PrototypeAndAttributes, subroutine
        attributes don't work properly for anonymous functions.

    *   This module is based on Parse::Keyword, which has a chronically broken
        implementation of closures. Kavorka uses PadWalker to attempt to work
        around the problem. This mostly seems to work, but you may experience
        some problems in edge cases, especially for anonymous functions and
        methods.

    *   If importing Kavorka's method modifiers into Moo/Mouse/Moose classes,
        pay attention to load order:

           use Moose;
           use Kavorka -all;   # ok

        If you do it this way, Moose's `before`, `after`, and `around`
        keywords will stomp on top of Kavorka's...

           use Kavorka -all;
           use Moose;          # STOMP, STOMP, STOMP!  :-(

        This can lead to delightfully hard to debug errors.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Kavorka>.

SUPPORT
    IRC: support is available through in the *#moops* channel on irc.perl.org
    <http://www.irc.perl.org/channels.html>.

SEE ALSO
    Kavorka::Manual.

    Inspirations: <http://perlcabal.org/syn/S06.html>, Function::Parameters,
    Method::Signatures.

    <http://en.wikipedia.org/wiki/The_Conversion_(Seinfeld)>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013-2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Kavorka
=======

Created:      2013-09-23
Home page:    <https://metacpan.org/release/Kavorka>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Kavorka>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.036	2014-11-01

 [ Bug Fixes ]
 - Avoid inadvertantly autovivifying things in Sub::Defer's big hash of
   deferred subs.
   Chris Tijerina++

0.035	2014-09-30

 [ Test Suite ]
 - Add a test case for Sub::Defer-related weirdness.
   <http://purl.org/NET/cpan-uri/rt/ticket/95786>

 [ Other ]
 - Make it easier to do weird things subclassing Kavorka::Multi.

0.034	2014-09-07

 - Avoid triggering RT#98666.
 - Updated: Switch from Sub::Name+Sub::Identify to the shiny new Sub::Util.

0.033	2014-08-22

 [ Bug Fixes ]
 - Type::Registry::DWIM changed its API. (Kavorka probably shouldn't be
   using it anyway because it's an undocumented internal part of
   Type::Utils.)
   SpiceMan++

0.032	2014-08-13

 [ Bug Fixes ]
 - Work around a bug in Moo(?) or maybe the Perl keyword API(?) by
   undeferring Moo's deferred subs.
   Fixes RT#95786.
   Jonathan Cast++
   <https://rt.cpan.org/Ticket/Display.html?id=95786>

0.031	2014-08-13

 [ Bug Fixes ]
 - Add `multi` to @EXPORT_OK.
   Thibaut Le Page++
 - Fix link to Kavorka::Manual::Signatures.
   Thibaut Le Page++

 [ Documentation ]
 - Update benchmarks given that Type::Tiny::XS is hellishly fast.

 [ Packaging ]
 - Recommend Type::Tiny::XS.

 [ Other ]
 - Updated: Bump required version of Return::Type.
 - Updated: Cope with changes to Exporter::Tiny; avoid triggering warnings.

0.030	2014-03-20

 [ Bug Fixes ]
 - Eliminate dependency on Devel::Pragma, which fixes Kavorka on threaded
   Perls.
   Fixes RT#94015.
   Piers Cawley++
   <https://rt.cpan.org/Ticket/Display.html?id=94015>

 [ Test Suite ]
 - Add test using Kavorka on threaded Perls.
   Aaron James Trevena++
   <http://purl.org/NET/cpan-uri/rt/ticket/94015>

 [ Other ]
 - Added: Introduce `f` as an alternative to `fun`.

0.029	2014-02-26

 - Added: Kavorka::TraitFor::Sub::begin.

0.028	2014-02-21	The 'perlsub-compat' Release

 [ Test Suite ]
 - Add test cases covering bare @ and % sigils.

 [ Other ]
 - Allow attributes to *precede* signatures.
 - Make parsing of parameter defaults laxer so that an equals sign followed
   by no expression is allowed.

0.027	2014-01-30

 - Minor change to the hook allowing keywords to deal with calls that
   bypass the Perl keyword API.

0.026	2014-01-30

 - Allow keywords to detect when the keyword sub has been called in a way
   that bypasses the Perl keyword API's custom parsing (e.g. using a
   prototype, or as a coderef). By default, keywords croak when this is
   detected.
 - Method modifier keywords can now detect when the keyword sub has been
   called bypassing custom parsing, and act sanely rather than croaking.

0.025	2013-12-23

 [ Bug Fixes ]
 - Value constraints for parameters without type constraints were being
   ignored.

 [ Test Suite ]
 - Tests for value constraints.

 [ Other ]
 - Added: Support non-block form of value constraints, taken from Perl 6 /
   Method::Signatures.

0.024	2013-12-18

 [ Bug Fixes ]
 - s/namespace::clean/namespace::sweep/

0.023	2013-12-17

 [ Bug Fixes ]
 - Fix test case - should not depend on Moops!
   Syohei Yoshida++
   <https://github.com/tobyink/p5-kavorka/issue/2>

0.022	2013-12-16

 [ Documentation ]
 - Fix typos in Kavorka::Manual::MultiSubs.
   Grzegorz Rożniecki++
   <https://github.com/tobyink/p5-kavorka/pull/1>

 [ Other ]
 - Added: Kavorka::TraitFor::Sub::fresh.
 - Added: Kavorka::TraitFor::Sub::override.

0.021	2013-11-30

 [ Documentation ]
 - Rearrange some of the traits documentation.

 [ Other ]
 - Added: Implement a `ref_alias` trait, inspired by Method::Signatures.
 - Disallow the `assumed` parameter trait for multi subs.
 - Split out most of the implementation for the `optional` trait (except
   for special parsing rules) into Kavorka::TraitFor::ReturnType::optional.
 - Split out some sanity_check stuff into the trait modules.

0.020	2013-11-27

 [ Documentation ]
 - Explain different ways to extend Kavorka.

 [ Other ]
 - Added: Kavorka::TraitFor::Parameter::assumed.
 - Can now use 'but' as another alias for 'is'/'does'.
 - Split out the assumed trait for return types as
   Kavorka::TraitFor::ReturnType::assumed.
 - Subs can now have traits.

0.019	2013-11-23

 - Added: Provide an `augment` keyword which will work in Moose/Mouse, but
   not Moo.
 - Added: Provide an `override` keyword which will work in Moose/Mouse, but
   not Moo.
 - Allow named parameters to be passed in as a hashref rather than a hash.
 - Split out some of the Kavorka::Parameter features (aliases, locked
   hashrefs, read-only parameters) into parameter traits.

0.018	2013-11-18

 - Allow non-Type::Tiny type constraints to be returned by type constraint
   expressions, provided they can be converted to Type::Tiny objects via
   Types::TypeTiny::to_TypeTiny.

0.017	2013-11-18

 - Extensibility via parameter traits.
 - Rename Kavorka::Signature::Parameter -> Kavorka::Parameter.
 - Rename Kavorka::Signature::ReturnType -> Kavorka::ReturnType.

0.016	2013-11-07

 [ Bug Fixes ]
 - Bypass the Internals::SvREADONLY prototype because handling of the ($;$)
   prototype changed in 5.16.0 and we were relying on the new behaviour.

0.015	2013-11-07

 [ BACK COMPAT ]
 - Change the syntax of lexical functions and methods to include the `my`
   keyword in their declaration.

 [ Bug Fixes ]
 - $sub->signature now returns undef if no signature was given in the
   declaration.
 - Most of the closure issues associated with Parse::Keyword are now worked
   around.

 [ Documentation ]
 - Document that fun ($x, $y=$x) {...} doesn't work how you might expect it
   to, and some work-arounds.

 [ Test Suite ]
 - Tidy up some of the closure tests.

 [ Other ]
 - Checking if @_ < 0 is silly; stop doing that.

0.014	2013-11-05	Gunpowder, treason and plot

 - Added: Lexical functions and methods.

Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "function signatures with the lure of the animal",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.023, CPAN::Meta::Converter version 2.142690",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Kavorka",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { suggests => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       recommends => { "Type::Tiny::XS" => "0.010" },
                                       requires   => {
                                                       "Data::Alias"      => 0,
                                                       "Exporter::Tiny"   => 0.026,
                                                       "match::simple"    => 0,
                                                       "Module::Runtime"  => 0,
                                                       "Moo"              => 1.003001,
                                                       "namespace::sweep" => 0.006,
                                                       "PadWalker"        => 0,
                                                       "Parse::Keyword"   => 0.06,
                                                       "perl"             => 5.014,
                                                       "Return::Type"     => 0.004,
                                                       "Sub::Util"        => 0,
                                                       "Type::Parser"     => 0.032,
                                                       "Type::Registry"   => 0.032,
                                                       "Type::Utils"      => 0.032,
                                                       "Types::Standard"  => 0.032,
                                                     },
                                     },
                        test      => {
                                       recommends => {
                                                       "Class::Method::Modifiers" => 2.03,
                                                       "Class::Tiny" => 0,
                                                       "Moose" => "2.0000",
                                                       "Mouse" => "1.00",
                                                       "Role::Tiny" => "1.003000",
                                                     },
                                       requires   => { "Test::Fatal" => 0, "Test::More" => 0.96, "Test::Requires" => 0 },
                                     },
                      },
  "provides"       => {
                        "Kavorka"                                   => { file => "lib/Kavorka.pm", version => 0.036 },
                        "Kavorka::MethodModifier"                   => { file => "lib/Kavorka/MethodModifier.pm", version => 0.036 },
                        "Kavorka::Multi"                            => { file => "lib/Kavorka/Multi.pm", version => 0.036 },
                        "Kavorka::Parameter"                        => { file => "lib/Kavorka/Parameter.pm", version => 0.036 },
                        "Kavorka::ReturnType"                       => { file => "lib/Kavorka/ReturnType.pm", version => 0.036 },
                        "Kavorka::Signature"                        => { file => "lib/Kavorka/Signature.pm", version => 0.036 },
                        "Kavorka::Sub"                              => { file => "lib/Kavorka/Sub.pm", version => 0.036 },
                        "Kavorka::Sub::After"                       => { file => "lib/Kavorka/Sub/After.pm", version => 0.036 },
                        "Kavorka::Sub::Around"                      => { file => "lib/Kavorka/Sub/Around.pm", version => 0.036 },
                        "Kavorka::Sub::Augment"                     => { file => "lib/Kavorka/Sub/Augment.pm", version => 0.036 },
                        "Kavorka::Sub::Before"                      => { file => "lib/Kavorka/Sub/Before.pm", version => 0.036 },
                        "Kavorka::Sub::ClassMethod"                 => { file => "lib/Kavorka/Sub/ClassMethod.pm", version => 0.036 },
                        "Kavorka::Sub::Fun"                         => { file => "lib/Kavorka/Sub/Fun.pm", version => 0.036 },
                        "Kavorka::Sub::Method"                      => { file => "lib/Kavorka/Sub/Method.pm", version => 0.036 },
                        "Kavorka::Sub::ObjectMethod"                => { file => "lib/Kavorka/Sub/ObjectMethod.pm", version => 0.036 },
                        "Kavorka::Sub::Override"                    => { file => "lib/Kavorka/Sub/Override.pm", version => 0.036 },
                        "Kavorka::TraitFor::Parameter::alias"       => { file => "lib/Kavorka/TraitFor/Parameter/alias.pm", version => 0.036 },
                        "Kavorka::TraitFor::Parameter::assumed"     => {
                                                                         file => "lib/Kavorka/TraitFor/Parameter/assumed.pm",
                                                                         version => 0.036,
                                                                       },
                        "Kavorka::TraitFor::Parameter::locked"      => {
                                                                         file => "lib/Kavorka/TraitFor/Parameter/locked.pm",
                                                                         version => 0.036,
                                                                       },
                        "Kavorka::TraitFor::Parameter::optional"    => {
                                                                         file => "lib/Kavorka/TraitFor/Parameter/optional.pm",
                                                                         version => 0.036,
                                                                       },
                        "Kavorka::TraitFor::Parameter::ref_alias"   => {
                                                                         file => "lib/Kavorka/TraitFor/Parameter/ref_alias.pm",
                                                                         version => 0.036,
                                                                       },
                        "Kavorka::TraitFor::Parameter::ro"          => { file => "lib/Kavorka/TraitFor/Parameter/ro.pm", version => 0.036 },
                        "Kavorka::TraitFor::ReturnType::assumed"    => {
                                                                         file => "lib/Kavorka/TraitFor/ReturnType/assumed.pm",
                                                                         version => 0.036,
                                                                       },
                        "Kavorka::TraitFor::Sub::begin"             => { file => "lib/Kavorka/TraitFor/Sub/begin.pm", version => 0.036 },
                        "Kavorka::TraitFor::Sub::fresh"             => { file => "lib/Kavorka/TraitFor/Sub/fresh.pm", version => 0.036 },
                        "Kavorka::TraitFor::Sub::override"          => { file => "lib/Kavorka/TraitFor/Sub/override.pm", version => 0.036 },
                        "MooseX::KavorkaInfo"                       => { file => "lib/MooseX/KavorkaInfo.pm", version => 0.036 },
                        "MooseX::KavorkaInfo::DummyInfo"            => { file => "lib/MooseX/KavorkaInfo.pm" },
                        "MooseX::KavorkaInfo::Trait::Method"        => { file => "lib/MooseX/KavorkaInfo.pm", version => 0.036 },
                        "MooseX::KavorkaInfo::Trait::WrappedMethod" => { file => "lib/MooseX/KavorkaInfo.pm", version => 0.036 },
                        "Parse::KeywordX"                           => { file => "lib/Parse/KeywordX.pm", version => 0.036 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Kavorka" },
                        homepage     => "https://metacpan.org/release/Kavorka",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-kavorka.git",
                                          web  => "https://github.com/tobyink/p5-kavorka",
                                        },
                        x_identifier => "http://purl.org/NET/cpan-uri/dist/Kavorka/project",
                        x_IRC        => "irc://irc.perl.org/#moops",
                      },
  "version"        => 0.036,
  "x_contributors" => [
                        "Grzegorz Ro\x{17C}niecki (XAERXESS) <xaerxess\@cpan.org>",
                        "Aaron James Trevena (TEEJAY) <teejay\@cpan.org>",
                        "Piers Cawley (PDCAWLEY) <pdcawley\@cpan.org>",
                        "Thibaut Le Page (THILP) <thilp\@cpan.org>",
                        "Chris Tijerina (CAMSPI) <camspi\@cpan.org>",
                      ],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

